### ISSUE TITLE
Virual MIDI device support on macOS

### ISSUE THREAD

**Issue by olgv** [keijiro/MinisVS#1]
Having the ability to select a different midi devices to listen from (then current), will allow bridging with MaxMSP over MIDI?

**Comment by keijiro** [comment]
Yes, that's a must-have feature for this add-on, but I'm pretty new to Bolt, so still I'm not sure what the best design is.

I'll think about it in a later phase of the development. Please stay tuned.

**Comment by keijiro** [comment]
I added the channel option to MIDI bolt units. Although you still can't select a specific device, you can workaround the problem by using different channel numbers in each device.

By the way, for bridging with MaxMSP, I think OSC is a better protocol choice. I'm thinking of implementing BoltOSC or something like that in the near future.

**Comment by olgv** [comment]
The channel addition is great!!!
I also found the `Midi Device Assigner` class and was trying to make use of it to point to a different device :P

I ended up testing with MaxMSP's MIDItester but somehow the MIDItester successfully talks to the controller (AKAI LPD8) and the controller successfully talks to Unity, but not the MaxMSP MIDItester with Unity. 
I saw mentions of the Unity Input Controls on the old Minis repo - I will have to look into that, might be why Unity is not seeing MaxMSP.

BoltOSC, that's another dream :)

**Comment by keijiro** [comment]
Are you using a virtual MIDI device driver to connect Max/MSP with Unity?

**Comment by olgv** [comment]
I am using the macOS Catalina's on-board one.

Here is my test setup:
![Screenshot 2020-11-05 at 09 28 55](https://user-images.githubusercontent.com/5272004/98224068-e8e08000-1f4a-11eb-93ff-0ce57eb4e8b7.png)
On channel `0` I get the LPD8 (OK), while on channel 10 I am expecting the ctrl'r value from MaxMSP.

And the virtual MIDI's ports look like this:
![Screenshot 2020-11-05 at 09 29 06](https://user-images.githubusercontent.com/5272004/98224207-1decd280-1f4b-11eb-95ad-1b78855e94cf.png)

Do you recommend a different virtual MIDI device?

**Comment by keijiro** [comment]
Thank you for the information. When I tried the IAC driver with Unity several years ago, it didn't work well. I'm not sure of the current status.

**Comment by olgv** [comment]
Ah! I see... good to know that the IAC Driver is not reliable with Unity.
Let me dig a bit around Unity and that and also check some alternatives.

**Comment by keijiro** [comment]
FYI, I released the OSC addon. I think it's handier for Max-Unity interoperability.  https://github.com/keijiro/BoltOscJack

**Comment by olgv** [comment]
Oh WOW!!! ..that's amasing! Thank you very much for approaching OSC and bringing it into Bolt.

The more I look in this MIDI issue, the more I realise that the problem lays in Unity not shaking hands with a virtual MIDI controller, at least on macOS. I tried two other virtual MIDI devices with no luck.

But now that you brought both MIDI and OSC into Bolt, is a good use case to keep them separate. **BoltMinis** can remain in use for MIDI hardware while **BoltOSCJack** becomes the bridge for all virtual ones.

**Comment by keijiro** [comment]
I modified the topic of this issue ticket. Although I can't fix it very soon, I keep it open for record purposes.
